File: dex/core/lockfile.py
"""Lock file management for Dex."""

from pathlib import Path

from dex.config.parser import load_lockfile, save_lockfile
from dex.config.schemas import AgentType, LockedPlugin, LockFile


class LockFileManager:
    """Manages the dex.lock file for deterministic installations."""

    def __init__(self, project_root: Path, agent: AgentType):
        """Initialize the lock file manager.

        Args:
            project_root: Path to the project root
            agent: Target agent platform
        """
        self._project_root = project_root
        self._agent = agent
        self._lockfile: LockFile | None = None
        self._modified = False

    def load(self) -> LockFile:
        """Load the lock file from disk.

        Creates a new empty lock file if one doesn't exist.

        Returns:
            The loaded or new lock file
        """
        self._lockfile = load_lockfile(self._project_root)
        if self._lockfile is None:
            self._lockfile = LockFile(agent=self._agent)
            self._modified = True
        return self._lockfile

    def save(self) -> None:
        """Save the lock file to disk if modified."""
        if self._lockfile is not None and self._modified:
            save_lockfile(self._project_root, self._lockfile)
            self._modified = False

    @property
    def lockfile(self) -> LockFile:
        """Get the current lock file, loading if necessary."""
        if self._lockfile is None:
            self.load()
        assert self._lockfile is not None
        return self._lockfile

    def get_locked_version(self, plugin_name: str) -> str | None:
        """Get the locked version for a plugin.

        Args:
            plugin_name: Name of the plugin

        Returns:
            Locked version string, or None if not locked
        """
        locked = self.lockfile.plugins.get(plugin_name)
        return locked.version if locked else None

    def get_locked_plugin(self, plugin_name: str) -> LockedPlugin | None:
        """Get the full locked plugin entry.

        Args:
            plugin_name: Name of the plugin

        Returns:
            LockedPlugin entry, or None if not locked
        """
        return self.lockfile.plugins.get(plugin_name)

    def lock_plugin(
        self,
        name: str,
        version: str,
        resolved_url: str,
        integrity: str | None = None,
        dependencies: dict[str, str] | None = None,
    ) -> None:
        """Lock a plugin to a specific version.

        Args:
            name: Plugin name
            version: Resolved version
            resolved_url: URL the package was resolved from
            integrity: Optional integrity hash
            dependencies: Optional dictionary of dependency versions
        """
        self.lockfile.plugins[name] = LockedPlugin(
            version=version,
            resolved=resolved_url,
            integrity=integrity or "",
            dependencies=dependencies or {},
        )
        self._modified = True

    def unlock_plugin(self, name: str) -> bool:
        """Remove a plugin from the lock file.

        Args:
            name: Plugin name to remove

        Returns:
            True if the plugin was removed, False if it wasn't locked
        """
        if name in self.lockfile.plugins:
            del self.lockfile.plugins[name]
            self._modified = True
            return True
        return False

    def is_locked(self, plugin_name: str) -> bool:
        """Check if a plugin is locked.

        Args:
            plugin_name: Name of the plugin

        Returns:
            True if the plugin has a lock entry
        """
        return plugin_name in self.lockfile.plugins

    def list_locked(self) -> list[str]:
        """List all locked plugin names.

        Returns:
            List of plugin names with lock entries
        """
        return list(self.lockfile.plugins.keys())

    def clear(self) -> None:
        """Clear all lock entries."""
        self.lockfile.plugins.clear()
        self._modified = True

File: internal/lockfile/lockfile.go
// Package lockfile provides lock file management for reproducible installs.
//
// The lock file is stored at dex.lock and pins exact versions of all installed
// plugins. This ensures that `dex install` produces identical results across
// different machines and times.
package lockfile

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
)

const (
	// LockFileVersion is the current lock file format version
	LockFileVersion = "1.0"

	// LockFileName is the lock file name
	LockFileName = "dex.lock"
)

// LockFile pins exact versions for reproducible installs.
// Stored at dex.lock (JSON format)
type LockFile struct {
	// Version is the lock file format version
	Version string `json:"version"`

	// Agent is the agentic platform (e.g., "claude-code", "cursor")
	Agent string `json:"agent"`

	// Plugins maps plugin names to their locked versions
	Plugins map[string]*LockedPlugin `json:"plugins"`

	// path is the path to the lock file (not serialized)
	path string
}

// LockedPlugin represents a locked plugin version.
type LockedPlugin struct {
	// Version is the exact version string
	Version string `json:"version"`

	// Resolved is the full URL or path used to fetch the plugin
	Resolved string `json:"resolved"`

	// Integrity is the content hash in format "sha256-{base64}"
	Integrity string `json:"integrity"`

	// Dependencies maps dependency plugin names to version constraints
	Dependencies map[string]string `json:"dependencies"`
}

// Load loads a lock file from the project root.
// Returns an empty lock file if the file doesn't exist.
func Load(projectRoot string) (*LockFile, error) {
	lockPath := filepath.Join(projectRoot, LockFileName)

	l := &LockFile{
		Version: LockFileVersion,
		Plugins: make(map[string]*LockedPlugin),
		path:    lockPath,
	}

	data, err := os.ReadFile(lockPath)
	if err != nil {
		if os.IsNotExist(err) {
			return l, nil
		}
		return nil, err
	}

	if err := json.Unmarshal(data, l); err != nil {
		return nil, err
	}

	// Ensure path is set after unmarshaling
	l.path = lockPath

	// Ensure Plugins map is initialized
	if l.Plugins == nil {
		l.Plugins = make(map[string]*LockedPlugin)
	}

	return l, nil
}

// Save writes the lock file to disk.
func (l *LockFile) Save() error {
	// Sort plugin entries for consistent output
	data, err := json.MarshalIndent(l, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(l.path, data, 0644)
}

// Get returns the locked version for a plugin (nil if not locked).
func (l *LockFile) Get(pluginName string) *LockedPlugin {
	return l.Plugins[pluginName]
}

// Set updates or adds a locked plugin.
func (l *LockFile) Set(pluginName string, locked *LockedPlugin) {
	if l.Plugins == nil {
		l.Plugins = make(map[string]*LockedPlugin)
	}

	// Ensure Dependencies map is initialized
	if locked.Dependencies == nil {
		locked.Dependencies = make(map[string]string)
	}

	l.Plugins[pluginName] = locked
}

// Remove removes a plugin from the lock file.
func (l *LockFile) Remove(pluginName string) {
	delete(l.Plugins, pluginName)
}

// Has checks if a plugin is in the lock file.
func (l *LockFile) Has(pluginName string) bool {
	_, exists := l.Plugins[pluginName]
	return exists
}

// LockedPlugins returns all locked plugin names (sorted).
func (l *LockFile) LockedPlugins() []string {
	names := make([]string, 0, len(l.Plugins))
	for name := range l.Plugins {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}
